BUG: special.chdtr: torch failures

gh-21294 fixed the delegation for `chdtr`, but has now surfaced some failures with PyTorch. cc @mdhaber 

```
___ test_support_alternative_backends[shapes1-float32-f_name_n_args4-torch] ____
scipy/special/tests/test_support_alternative_backends.py:100: in test_support_alternative_backends
    xp_assert_close(res, ref, atol=10*eps)
E   AssertionError: Scalars are not close!
E   
E   Expected nan but got 0.0.
E   Absolute difference: nan (up to 1.1920928955078125e-06 allowed)
E   Relative difference: nan (up to 0.0013810679320049757 allowed)
        args_np    = [array(-0.1832966, dtype=float32), array(-0.6893303, dtype=float32)]
        args_xp    = [tensor(-0.1833), tensor(-0.6893)]
        dtype      = 'float32'
        dtype_np   = <class 'numpy.float32'>
        dtype_xp   = torch.float32
        eps        = np.float32(1.1920929e-07)
        f          = <function support_alternative_backends.<locals>.wrapped at 0x7f9c81998360>
        f_name     = 'chdtr'
        f_name_n_args = ('chdtr', 2)
        n_args     = 2
        ref        = tensor(nan)
        res        = tensor(0.)
        rng        = Generator(PCG64) at 0x7F9C038A9C40
        shapes     = [(), ()]
        xp         = <module 'torch' from '/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/torch/__init__.py'>
_ test_support_alternative_backends[shapes1-float32-f_name_n_args4-jax.numpy] __
scipy/special/tests/test_support_alternative_backends.py:100: in test_support_alternative_backends
    xp_assert_close(res, ref, atol=10*eps)
        args_np    = [array(-0.1832966, dtype=float32), array(-0.6893303, dtype=float32)]
        args_xp    = [Array(-0.1832966, dtype=float32), Array(-0.6893303, dtype=float32)]
        dtype      = 'float32'
        dtype_np   = <class 'numpy.float32'>
        dtype_xp   = <class 'jax.numpy.float32'>
        eps        = np.float32(1.1920929e-07)
        f          = <function support_alternative_backends.<locals>.wrapped at 0x7f9c81998360>
        f_name     = 'chdtr'
        f_name_n_args = ('chdtr', 2)
        n_args     = 2
        ref        = Array(nan, dtype=float32)
        res        = Array(0., dtype=float32)
        rng        = Generator(PCG64) at 0x7F9C00DFDC40
        shapes     = [(), ()]
        xp         = <module 'jax.numpy' from '/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/jax/numpy/__init__.py'>
/opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/contextlib.py:81: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=0.00138107, atol=1.19209e-06
E   
E   nan location mismatch:
E    ACTUAL: array(0., dtype=float32)
E    DESIRED: array(nan, dtype=float32)
        args       = (<function assert_allclose.<locals>.compare at 0x7f9c00e85120>, array(0., dtype=float32), array(nan, dtype=float32))
        func       = <function assert_array_compare at 0x7f9c81954860>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=0.00138107, atol=1.19209e-06', 'strict': False, ...}
        self       = <contextlib._GeneratorContextManager object at 0x7f9c81949b90>
___ test_support_alternative_backends[shapes2-float32-f_name_n_args4-torch] ____
 DESIRED: array([     nan,      nan, 0.680044,      nan,      nan,      nan,
       0.670524,      nan,      nan, 0.821572], dtype=float32)
FAILED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[shapes2-float64-f_name_n_args4-torch] - AssertionError: Tensor-likes are not close!

Mismatched elements: 8 / 10 (80.0%)
Greatest absolute difference: nan at index (0,) (up to 2.220446049250313e-15 allowed)
Greatest relative difference: nan at index (0,) (up to 5.96046447[753](https://github.com/scipy/scipy/actions/runs/10229735671/job/28303726377?pr=21304#step:11:754)9063e-08 allowed)
FAILED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[shapes2-float64-f_name_n_args4-jax.numpy] - AssertionError: 
Not equal to tolerance rtol=5.96046e-08, atol=2.22045e-15

nan location mismatch:
 ACTUAL: array([0.      , 0.      , 0.      , 0.773903, 0.      , 0.      ,
       0.158065, 0.      , 0.      , 0.      ])
 DESIRED: array([     nan,      nan,      nan, 0.773903,      nan,      nan,
       0.158065,      nan,      nan,      nan])
FAILED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[shapes3-float32-f_name_n_args4-torch] - AssertionError: Tensor-likes are not close!

Mismatched elements: 30 / 110 (27.3%)
Greatest absolute difference: nan at index (4, 0) (up to 1.1920928955078125e-06 allowed)
Greatest relative difference: nan at index (4, 0) (up to 0.0013810679320049[757](https://github.com/scipy/scipy/actions/runs/10229735671/job/28303726377?pr=21304#step:11:758) allowed)
FAILED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[shapes3-float32-f_name_n_args4-jax.numpy] - AssertionError: 
Not equal to tolerance rtol=0.00138107, atol=1.19209e-06

nan location mismatch:
 ACTUAL: array([[     nan,      nan, 0.573669,      nan,      nan,      nan,
        0.40[764](https://github.com/scipy/scipy/actions/runs/10229735671/job/28303726377?pr=21304#step:11:765)7,      nan,      nan, 0.812421],
       [     nan,      nan, 0.829262,      nan,      nan,      nan,...
 DESIRED: array([[     nan,      nan, 0.573669,      nan,      nan,      nan,
        0.407647,      nan,      nan, 0.812421],
       [     nan,      nan, 0.829262,      nan,      nan,      nan,...
FAILED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[shapes3-float64-f_name_n_args4-torch] - AssertionError: Tensor-likes are not close!

Mismatched elements: 80 / 110 (72.7%)
Greatest absolute difference: nan at index (0, 0) (up to 2.220446049250313e-15 allowed)
Greatest relative difference: nan at index (0, 0) (up to 5.9604644[775](https://github.com/scipy/scipy/actions/runs/10229735671/job/28303726377?pr=21304#step:11:776)39063e-08 allowed)
FAILED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[shapes3-float64-f_name_n_args4-jax.numpy] - AssertionError: 
Not equal to tolerance rtol=5.96046e-08, atol=2.22045e-15

nan location mismatch:
 ACTUAL: array([[0.      , 0.      , 0.      , 0.      , 0.      , 0.      ,
        0.      , 0.      , 0.      , 0.      ],
       [0.      , 0.      , 0.      , 0.      , 0.      , 0.      ,...
 DESIRED: array([[     nan,      nan,      nan,      nan,      nan,      nan,
             nan,      nan,      nan,      nan],
       [     nan,      nan,      nan,      nan,      nan,      nan,...
= 10 failed, 35673 passed, 9410 skipped, 684 deselected, 2225 xfailed, 1 xpassed in 486.91s (0:08:06) =
```

--- a/scipy/special/_support_alternative_backends.py
+++ b/scipy/special/_support_alternative_backends.py
@@ -90,10 +90,12 @@ def _chdtr(xp, spx):
         return None
 
     def __chdtr(v, x):
-        res = xp.where(x >= 0, gammainc(v/2, x/2), 0)
-        i_nan = ((x == 0) & (v == 0)) | xp.isnan(x) | xp.isnan(v)
-        res = xp.where(i_nan, xp.nan, res)
-        return res
+        res = gammainc(v / 2, x / 2)  # this is almost all we need
+        # The rest can be removed when google/jax#20507 is resolved
+        mask = (v == 0) & (x > 0)  # JAX returns NaN
+        res = xp.where(mask, 1., res)
+        mask = xp.isinf(v) & xp.isinf(x)  # JAX returns 1.0
+        return xp.where(mask, xp.nan, res)
     return __chdtr
 
 

--- a/scipy/special/tests/test_support_alternative_backends.py
+++ b/scipy/special/tests/test_support_alternative_backends.py
@@ -98,3 +98,14 @@ def test_support_alternative_backends(xp, f_name_n_args, dtype, shapes):
 
     eps = np.finfo(dtype_np).eps
     xp_assert_close(res, ref, atol=10*eps)
+
+
+@array_api_compatible
+def test_chdtr_gh21311(xp):
+    # the edge case behavior of generic chdtr was not right; see gh-21311
+    # make to test at least these cases
+    x = np.asarray([-np.inf, -1., 0., 1., np.inf])
+    v = x.reshape(-1, 1)
+    ref = special.chdtr(v, x)
+    res = special.chdtr(xp.asarray(v), xp.asarray(x))
+    xp_assert_close(res, xp.asarray(ref))

--- a/scipy/special/tests/test_support_alternative_backends.py
+++ b/scipy/special/tests/test_support_alternative_backends.py
@@ -103,7 +103,8 @@ def test_support_alternative_backends(xp, f_name_n_args, dtype, shapes):
 @array_api_compatible
 def test_chdtr_gh21311(xp):
     # the edge case behavior of generic chdtr was not right; see gh-21311
-    # make to test at least these cases
+    # be sure to test at least these cases
+    # should add `np.nan` into the mix when gh-21317 is resolved
     x = np.asarray([-np.inf, -1., 0., 1., np.inf])
     v = x.reshape(-1, 1)
     ref = special.chdtr(v, x)

